### DISCUSSION TITLE
Difference between `InstanceState.unloaded` and `InstanceState.unloaded_expirable`

### DISCUSSION THREAD

**Discussion by provinzkraut** [sqlalchemy/sqlalchemy#9913] [Usage Questions] [answered]
I was looking at the documentation for `InstanceState.unloaded_expirable` and noticed it's exactly the same as `InstanceState.unloaded`.

Checking the code, `unloaded_expirable` seems to just return `unloaded` unchanged:

https://github.com/sqlalchemy/sqlalchemy/blob/5f26a66aeca841ca2ec714d3473394d61a29ed9a/lib/sqlalchemy/orm/state.py#L823-L845

So I'm wondering, is there a difference? Or this this perhaps simply done for reasons of backwards compatibility, and their behaviour used to differ?

**Comment by zzzeek** [comment, marked as answer]
doing a git blame, it was added in 5abb036e9b9eba0f61cf9 as a subset of .unloaded, those which also included the "expire_missing" flag on the attribute.   this seems to have indicated attributes that are unloaded, but also should be "expired" when an expire() operation for the whole object is called.

Then, in 33119301bb3efa143 , the "expire_missing" flag was renamed to "load_on_unexpire", but was removed from the definition of unloaded_expired to be a mirror of .unloaded, as you suspected.   This indicates that an expire() operation can now be applied to all unloaded attributes.   I would assume that in this rev, the behavior of deferred attributes was enhanced so that they can accommodate being expired while still doing whatever it is we want them to do during an unexpire.

one impression I am getting pretty strongly is that in 5abb036e9b9eb this definitely should not have been a public attribute.   not sure what I was thinking on that, as this is definitely implementation specific.

**Reply by provinzkraut** [reply]
Thanks a lot for the clarification! I'll keep using `.unloaded` then and don't worry about the other one.

Should this maybe be mentioned in the docs somewhere?

**Reply by zzzeek** [reply]
im deprecating it now
